src/web/mod.rs

```diff
@@ -335,24 +343,53 @@ fn latest_version(versions_json: &Vec<String>, req_version: &str) -> Option<Stri
 }
 
 
+pub struct Server {
+    inner: Listening,
+}
 
-/// Starts cratesfyi web server
-pub fn start_web_server(sock_addr: Option<&str>) {
-    // poke all the metrics counters to instantiate and register them
-    metrics::TOTAL_BUILDS.inc_by(0);
-    metrics::SUCCESSFUL_BUILDS.inc_by(0);
-    metrics::FAILED_BUILDS.inc_by(0);
-    metrics::NON_LIBRARY_BUILDS.inc_by(0);
-    metrics::UPLOADED_FILES_TOTAL.inc_by(0);
+impl Server {
+    pub fn start(addr: Option<&str>) -> Self {
+        let server = Self::start_inner(addr.unwrap_or(DEFAULT_BIND), Box::new(|| Pool::new()));
+        info!("Running docs.rs web server on http://{}", server.addr());
+        server
+    }
 
-    let cratesfyi = CratesfyiHandler::new();
-    let addr = sock_addr.unwrap_or("0.0.0.0:3000");
-    // need to assign this so it's not dropped before the function ends
-    let _server = Iron::new(cratesfyi).http(addr)
-        .unwrap_or_else(|_| panic!("Failed to bind to socket on {}", addr));
-    info!("Running docs.rs web server on http://{}", addr);
-}
+    #[cfg(test)]
+    pub(crate) fn start_test(conn: Arc<Mutex<Connection>>) -> Self {
+        Self::start_inner("127.0.0.1:0", Box::new(move || Pool::new_simple(conn.clone())))
+    }
+
+    fn start_inner(addr: &str, pool_factory: Box<dyn Fn() -> Pool + Send + Sync>) -> Self {
+        // poke all the metrics counters to instantiate and register them
+        metrics::TOTAL_BUILDS.inc_by(0);
+        metrics::SUCCESSFUL_BUILDS.inc_by(0);
+        metrics::FAILED_BUILDS.inc_by(0);
+        metrics::NON_LIBRARY_BUILDS.inc_by(0);
+        metrics::UPLOADED_FILES_TOTAL.inc_by(0);
 
+        let cratesfyi = CratesfyiHandler::new(pool_factory);
+        let inner = Iron::new(cratesfyi).http(addr)
+            .unwrap_or_else(|_| panic!("Failed to bind to socket on {}", addr));
+
+        Server { inner }
+    }
+
+    pub(crate) fn addr(&self) -> SocketAddr {
+        self.inner.socket
+    }
+
+    /// Iron is bugged, and it never closes the server even when the listener is dropped. To
+    /// avoid never-ending tests this method forgets about the server, leaking it and allowing the
+    /// program to end.
+    ///
+    /// The OS will then close all the dangling servers once the process exits.
+    ///
+    /// https://docs.rs/iron/0.6.1/iron/struct.Listening.html#method.close
```

Comment:
nit: we're still on 0.5 :(

This is [fixed in hyper 0.11](https://github.com/hyperium/hyper/issues/338#issuecomment-347044047), but even the latest version of iron [depends on 0.10](https://docs.rs/crate/iron/0.6.1/source/Cargo.toml.orig).

```suggestion
    /// https://docs.rs/iron/0.5.1/iron/struct.Listening.html#method.close
```